app/src/main/java/com/itsaky/androidide/fragments/ProjectListFragment.kt

```diff
@@ -0,0 +1,140 @@
+import com.itsaky.androidide.R
+import com.itsaky.androidide.activities.MainActivity
+import com.itsaky.androidide.adapters.ProjectListAdapter
+import com.itsaky.androidide.databinding.FragmentProjectListBinding
+import com.itsaky.androidide.models.ProjectInfoDetails
+import com.itsaky.androidide.models.toProjectInfoDetails
+import com.itsaky.androidide.provider.IDEViewModelProvider
+import com.itsaky.androidide.utils.ProjectInfoDetailsUtils
+import com.itsaky.androidide.viewmodel.MainViewModel
+import com.itsaky.androidide.viewmodel.ProjectInfoViewModel
+import java.io.File
+
+
+class ProjectListFragment :
+  FragmentWithBinding<FragmentProjectListBinding>(R.layout.fragment_project_list,
+    FragmentProjectListBinding::bind) {
+
+  private var adapter: ProjectListAdapter? = null
+
+  private val viewModel by viewModels<MainViewModel>(ownerProducer = { requireActivity() })
+  private val projectInfoViewModel: ProjectInfoViewModel by viewModels { IDEViewModelProvider.Factory }
+
+  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+    super.onViewCreated(view, savedInstanceState)
+
+    setupRecyclerView()
+    observeProjectInfoList()
+
+    projectInfoViewModel.getAllProjectInfo()
+
+    binding.exitButton.setOnClickListener {
+      viewModel.setScreen(MainViewModel.SCREEN_MAIN)
+    }
+
+  }
+
+  private fun setupRecyclerView() {
+    val layoutManager = LinearLayoutManager(requireContext())
+    layoutManager.orientation = LinearLayoutManager.VERTICAL
+    binding.list.layoutManager = layoutManager
+
+  }
+
+  private fun observeProjectInfoList() {
+    projectInfoViewModel.projectInfoList.observe(viewLifecycleOwner) {
+      val projectList = it.map { projectInfo ->
+        projectInfo.toProjectInfoDetails()
+      }
+
+      onFileListChanged(projectList)
+    }
+
+  }
+
+  private fun onFileListChanged(projectList: List<ProjectInfoDetails>) {
+    val sortOptions = ProjectInfoDetailsUtils.ProjectSortOptions(
+      ProjectInfoDetailsUtils.SortBy.OPEN_LAST, ProjectInfoDetailsUtils.Order.ASCENDING)
+      .createComparator()
+    val sortedProjectList = projectList.sortedWith(sortOptions)
+
+    adapter = ProjectListAdapter(sortedProjectList) { project, _ ->
+      openProject(project.file)
+    }
+
+    binding.list.adapter = adapter
+    setupNoProjectList(sortedProjectList.isEmpty())
+  }
+
+  private fun setupNoProjectList(isEmpty: Boolean) {
+    binding.list.visibility = if (isEmpty) View.GONE else View.VISIBLE
+    binding.noOpenProjectsSummary.visibility = if (isEmpty) View.VISIBLE else View.GONE
+
+    binding.noOpenProjectsSummary.movementMethod = LinkMovementMethod()
+    val openExistingProjectSpan: ClickableSpan = object : ClickableSpan() {
+      override fun onClick(widget: View) {
+        pickDirectory()
+      }
+    }
+
+    val sb = SpannableStringBuilder()
+    appendClickableSpan(sb, R.string.msg_empty_recent_projects, openExistingProjectSpan)
+    binding.noOpenProjectsSummary.text = sb
+  }
```

Comment:
You don't need to implement this manually. Simply extend this fragment with [`EmptyStateFragment`](https://github.com/AndroidIDEOfficial/AndroidIDE/blob/dev/app/src/main/java/com/itsaky/androidide/fragments/EmptyStateFragment.kt) and set the message.

There is also [`RecyclerViewFragment`](https://github.com/AndroidIDEOfficial/AndroidIDE/blob/dev/app/src/main/java/com/itsaky/androidide/fragments/RecyclerViewFragment.kt) which is useful for fragments that only show a `RecyclerView`. It handles the empty state as well.